// atozas-react-auth-kit/src/api/client.ts
import axios, { AxiosInstance, AxiosError, InternalAxiosRequestConfig } from 'axios';

let refreshTokenPromise: Promise<string | null> | null = null;

export function createApiClient(
  baseURL: string,
  getAccessToken: () => string | null,
  setAccessToken: (token: string | null) => void,
  onRefreshToken: () => Promise<string | null>,
  onAuthError?: (error: any) => void
): AxiosInstance {
  const client = axios.create({
    baseURL,
    withCredentials: true, // Important for cookies
    headers: {
      'Content-Type': 'application/json',
    },
  });

  // Request interceptor - add access token
  client.interceptors.request.use(
    (config: InternalAxiosRequestConfig) => {
      const token = getAccessToken();
      if (token) {
        config.headers.Authorization = `Bearer ${token}`;
      }
      return config;
    },
    (error) => Promise.reject(error)
  );

  // Response interceptor - handle 401 and refresh token
  client.interceptors.response.use(
    (response) => response,
    async (error: AxiosError) => {
      const originalRequest = error.config as InternalAxiosRequestConfig & { _retry?: boolean };

      // If 401 and not already retrying
      if (error.response?.status === 401 && !originalRequest._retry) {
        originalRequest._retry = true;

        try {
          // Prevent multiple simultaneous refresh requests
          if (!refreshTokenPromise) {
            refreshTokenPromise = onRefreshToken().finally(() => {
              refreshTokenPromise = null;
            });
          }

          const newAccessToken = await refreshTokenPromise;

          if (newAccessToken) {
            setAccessToken(newAccessToken);
            originalRequest.headers.Authorization = `Bearer ${newAccessToken}`;
            return client(originalRequest);
          }
        } catch (refreshError) {
          // Refresh failed
          setAccessToken(null);
          if (onAuthError) {
            onAuthError(refreshError);
          }
          return Promise.reject(refreshError);
        }
      }

      return Promise.reject(error);
    }
  );

  return client;
}
