// atozas-auth-kit-express/src/router.ts
import { Router, Request, Response } from 'express';
import { OAuth2Client } from 'google-auth-library';
import { AuthConfig, AuthRequest } from './types';
import User from './models/User';
import Otp from './models/Otp';
import { generateAccessToken, generateRefreshToken, verifyRefreshToken } from './utils/tokens';
import { generateOtp, hashOtp, compareOtp } from './utils/otp';
import { checkRateLimit } from './utils/rateLimit';
import { createRequireAuth } from './middleware/requireAuth';

export function createAuthRouter(config: AuthConfig): Router {
  const router = Router();
  const googleClient = new OAuth2Client(config.googleClientId);
  
  // Default values
  const accessTokenExpiry = config.accessTokenExpiry || '15m';
  const refreshTokenExpiry = config.refreshTokenExpiry || '30d';
  const otpExpiry = config.otpExpiry || 5; // minutes
  const otpLength = config.otpLength || 6;
  const otpRateLimit = config.otpRateLimit || {
    maxAttempts: 3,
    windowMs: 15 * 60 * 1000,
  };
  
  const cookieOptions = {
    httpOnly: true,
    secure: process.env.NODE_ENV === 'production',
    sameSite: 'lax' as const,
    maxAge: 30 * 24 * 60 * 60 * 1000, // 30 days
    ...config.cookieOptions,
  };
  
  // Helper to set tokens
  const setAuthTokens = (res: Response, userId: string, email: string, provider: 'google' | 'email') => {
    const accessToken = generateAccessToken(
      { userId, email, provider },
      config.jwtSecret,
      accessTokenExpiry
    );
    
    const refreshToken = generateRefreshToken(
      { userId, email, provider },
      config.jwtRefreshSecret,
      refreshTokenExpiry
    );
    
    res.cookie('refreshToken', refreshToken, cookieOptions);
    
    return { accessToken, refreshToken };
  };
  
  // POST /google - Google OAuth login
  router.post('/google', async (req: Request, res: Response) => {
    try {
      const { credential } = req.body;
      
      if (!credential) {
        return res.status(400).json({ error: 'Credential is required' });
      }
      
      // Verify the Google ID token
      const ticket = await googleClient.verifyIdToken({
        idToken: credential,
        audience: config.googleClientId,
      });
      
      const payload = ticket.getPayload();
      
      if (!payload || !payload.email) {
        return res.status(400).json({ error: 'Invalid token payload' });
      }
      
      const { email, name, picture, sub: googleId } = payload;
      
      // Find or create user
      let user = await User.findOne({ email });
      
      if (!user) {
        user = await User.create({
          email,
          name,
          picture,
          provider: 'google',
          googleId,
        });
      } else if (user.provider === 'email') {
        // Link Google account to existing email user
        user.googleId = googleId;
        user.name = user.name || name;
        user.picture = user.picture || picture;
        await user.save();
      }
      
      const { accessToken } = setAuthTokens(res, user._id.toString(), user.email, 'google');
      
      res.json({
        success: true,
        accessToken,
        user: {
          id: user._id.toString(),
          email: user.email,
          name: user.name,
          picture: user.picture,
          provider: user.provider,
        },
      });
    } catch (error: any) {
      console.error('Google auth error:', error);
      res.status(401).json({ error: 'Google authentication failed', details: error.message });
    }
  });
  
  // POST /email/send-otp - Send OTP to email
  router.post('/email/send-otp', async (req: Request, res: Response) => {
    try {
      const { email } = req.body;
      
      if (!email || !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email)) {
        return res.status(400).json({ error: 'Valid email is required' });
      }
      
      const ipAddress = req.ip || req.socket.remoteAddress || 'unknown';
      const rateLimitKey = `otp:${email.toLowerCase()}:${ipAddress}`;
      
      // Check rate limit
      const rateLimitResult = checkRateLimit(
        rateLimitKey,
        otpRateLimit.maxAttempts,
        otpRateLimit.windowMs
      );
      
      if (!rateLimitResult.allowed) {
        return res.status(429).json({
          error: 'Too many OTP requests',
          retryAfter: Math.ceil((rateLimitResult.resetAt - Date.now()) / 1000),
        });
      }
      
      // Generate OTP
      const otp = generateOtp(otpLength);
      const hashedOtp = await hashOtp(otp);
      
      // Save OTP to database
      const expiresAt = new Date(Date.now() + otpExpiry * 60 * 1000);
      await Otp.create({
        email: email.toLowerCase(),
        otp: hashedOtp,
        expiresAt,
        used: false,
        ipAddress,
      });
      
      // Send OTP via email
      try {
        await config.sendEmailOtp(email, otp);
      } catch (error: any) {
        console.error('Failed to send OTP email:', error);
        return res.status(500).json({ error: 'Failed to send OTP email' });
      }
      
      res.json({
        success: true,
        message: 'OTP sent to email',
        expiresIn: otpExpiry * 60, // in seconds
      });
    } catch (error: any) {
      console.error('Send OTP error:', error);
      res.status(500).json({ error: 'Failed to send OTP' });
    }
  });
  
  // POST /email/verify-otp - Verify OTP and login
  router.post('/email/verify-otp', async (req: Request, res: Response) => {
    try {
      const { email, otp } = req.body;
      
      if (!email || !otp) {
        return res.status(400).json({ error: 'Email and OTP are required' });
      }
      
      // Find valid OTP
      const otpDoc = await Otp.findOne({
        email: email.toLowerCase(),
        used: false,
        expiresAt: { $gt: new Date() },
      }).sort({ createdAt: -1 });
      
      if (!otpDoc) {
        return res.status(400).json({ error: 'Invalid or expired OTP' });
      }
      
      // Verify OTP
      const isValid = await compareOtp(otp, otpDoc.otp);
      
      if (!isValid) {
        return res.status(400).json({ error: 'Invalid OTP' });
      }
      
      // Mark OTP as used
      otpDoc.used = true;
      await otpDoc.save();
      
      // Find or create user
      let user = await User.findOne({ email: email.toLowerCase() });
      
      if (!user) {
        user = await User.create({
          email: email.toLowerCase(),
          provider: 'email',
        });
      }
      
      const { accessToken } = setAuthTokens(res, user._id.toString(), user.email, 'email');
      
      res.json({
        success: true,
        accessToken,
        user: {
          id: user._id.toString(),
          email: user.email,
          name: user.name,
          picture: user.picture,
          provider: user.provider,
        },
      });
    } catch (error: any) {
      console.error('Verify OTP error:', error);
      res.status(500).json({ error: 'Failed to verify OTP' });
    }
  });
  
  // POST /refresh - Refresh access token
  router.post('/refresh', async (req: Request, res: Response) => {
    try {
      const refreshToken = req.cookies.refreshToken;
      
      if (!refreshToken) {
        return res.status(401).json({ error: 'No refresh token provided' });
      }
      
      try {
        const decoded = verifyRefreshToken(refreshToken, config.jwtRefreshSecret);
        
        // Verify user still exists
        const user = await User.findById(decoded.userId);
        
        if (!user) {
          return res.status(401).json({ error: 'User not found' });
        }
        
        // Generate new access token
        const accessToken = generateAccessToken(
          { userId: user._id.toString(), email: user.email, provider: user.provider },
          config.jwtSecret,
          accessTokenExpiry
        );
        
        res.json({
          success: true,
          accessToken,
        });
      } catch (error) {
        return res.status(401).json({ error: 'Invalid or expired refresh token' });
      }
    } catch (error: any) {
      console.error('Refresh token error:', error);
      res.status(500).json({ error: 'Failed to refresh token' });
    }
  });
  
  // GET /me - Get current user
  router.get('/me', createRequireAuth(config), async (req: AuthRequest, res: Response) => {
    try {
      if (!req.user) {
        return res.status(401).json({ error: 'Not authenticated' });
      }
      
      res.json({
        success: true,
        user: req.user,
      });
    } catch (error: any) {
      console.error('Get user error:', error);
      res.status(500).json({ error: 'Failed to get user' });
    }
  });
  
  // POST /logout - Logout user
  router.post('/logout', (req: Request, res: Response) => {
    res.clearCookie('refreshToken', cookieOptions);
    res.json({
      success: true,
      message: 'Logged out successfully',
    });
  });
  
  return router;
}
