// atozas-react-auth-kit/src/context/AuthContext.tsx
import React, { createContext, useState, useEffect, useCallback, useMemo } from 'react';
import { GoogleOAuthProvider } from '@react-oauth/google';
import { AuthContextValue, AuthProviderProps, User } from '../types';
import { createApiClient } from '../api/client';
import { AxiosInstance } from 'axios';

export const AuthContext = createContext<AuthContextValue | undefined>(undefined);

// Export the API client instance for components to use
export let apiClient: AxiosInstance;

export function AuthProvider({
  children,
  apiUrl,
  googleClientId,
  enableLocalStorage = false,
  onAuthError,
}: AuthProviderProps) {
  const [user, setUser] = useState<User | null>(null);
  const [accessToken, setAccessTokenState] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  // Token storage helper
  const setAccessToken = useCallback((token: string | null) => {
    setAccessTokenState(token);
    if (enableLocalStorage) {
      if (token) {
        localStorage.setItem('accessToken', token);
      } else {
        localStorage.removeItem('accessToken');
        localStorage.removeItem('user');
      }
    }
  }, [enableLocalStorage]);

  const getAccessToken = useCallback(() => {
    if (accessToken) return accessToken;
    if (enableLocalStorage) {
      return localStorage.getItem('accessToken');
    }
    return null;
  }, [accessToken, enableLocalStorage]);

  // Refresh token function
  const refreshToken = useCallback(async (): Promise<string | null> => {
    try {
      const response = await fetch(`${apiUrl}/refresh`, {
        method: 'POST',
        credentials: 'include',
        headers: {
          'Content-Type': 'application/json',
        },
      });

      if (!response.ok) {
        throw new Error('Failed to refresh token');
      }

      const data = await response.json();
      if (data.accessToken) {
        setAccessToken(data.accessToken);
        return data.accessToken;
      }
      return null;
    } catch (error) {
      console.error('Token refresh failed:', error);
      return null;
    }
  }, [apiUrl, setAccessToken]);

  // Initialize API client
  useEffect(() => {
    apiClient = createApiClient(
      apiUrl,
      getAccessToken,
      setAccessToken,
      refreshToken,
      onAuthError
    );
  }, [apiUrl, getAccessToken, setAccessToken, refreshToken, onAuthError]);

  // Login function
  const login = useCallback((token: string, userData: User) => {
    setAccessToken(token);
    setUser(userData);
    if (enableLocalStorage) {
      localStorage.setItem('user', JSON.stringify(userData));
    }
  }, [enableLocalStorage, setAccessToken]);

  // Logout function
  const logout = useCallback(async () => {
    try {
      await apiClient.post('/logout');
    } catch (error) {
      console.error('Logout error:', error);
    } finally {
      setAccessToken(null);
      setUser(null);
    }
  }, [setAccessToken]);

  // Initialize auth state on mount
  useEffect(() => {
    const initAuth = async () => {
      try {
        const token = enableLocalStorage ? localStorage.getItem('accessToken') : null;
        const storedUser = enableLocalStorage ? localStorage.getItem('user') : null;

        if (token && storedUser) {
          setAccessToken(token);
          setUser(JSON.parse(storedUser));

          // Verify token is still valid by fetching user
          try {
            const response = await fetch(`${apiUrl}/me`, {
              headers: {
                Authorization: `Bearer ${token}`,
              },
              credentials: 'include',
            });

            if (response.ok) {
              const data = await response.json();
              setUser(data.user);
            } else {
              // Token invalid, try to refresh
              const newToken = await refreshToken();
              if (newToken) {
                const retryResponse = await fetch(`${apiUrl}/me`, {
                  headers: {
                    Authorization: `Bearer ${newToken}`,
                  },
                  credentials: 'include',
                });
                if (retryResponse.ok) {
                  const data = await retryResponse.json();
                  setUser(data.user);
                } else {
                  setAccessToken(null);
                  setUser(null);
                }
              } else {
                setAccessToken(null);
                setUser(null);
              }
            }
          } catch (error) {
            console.error('Failed to verify auth:', error);
            setAccessToken(null);
            setUser(null);
          }
        }
      } catch (error) {
        console.error('Auth initialization error:', error);
      } finally {
        setIsLoading(false);
      }
    };

    initAuth();
  }, [apiUrl, enableLocalStorage, refreshToken, setAccessToken]);

  const contextValue = useMemo<AuthContextValue>(() => ({
    user,
    isAuthenticated: !!user,
    isLoading,
    accessToken,
    login,
    logout,
    refreshToken,
  }), [user, isLoading, accessToken, login, logout, refreshToken]);

  return (
    <GoogleOAuthProvider clientId={googleClientId}>
      <AuthContext.Provider value={contextValue}>
        {children}
      </AuthContext.Provider>
    </GoogleOAuthProvider>
  );
}
