// atozas-react-auth-kit/dist/index.mjs
"use client";

// src/context/AuthContext.tsx
import { createContext, useState, useEffect, useCallback, useMemo } from "react";
import { GoogleOAuthProvider } from "@react-oauth/google";

// src/api/client.ts
import axios from "axios";
var refreshTokenPromise = null;
function createApiClient(baseURL, getAccessToken, setAccessToken, onRefreshToken, onAuthError) {
  const client = axios.create({
    baseURL,
    withCredentials: true,
    // Important for cookies
    headers: {
      "Content-Type": "application/json"
    }
  });
  client.interceptors.request.use(
    (config) => {
      const token = getAccessToken();
      if (token) {
        config.headers.Authorization = `Bearer ${token}`;
      }
      return config;
    },
    (error) => Promise.reject(error)
  );
  client.interceptors.response.use(
    (response) => response,
    async (error) => {
      const originalRequest = error.config;
      if (error.response?.status === 401 && !originalRequest._retry) {
        originalRequest._retry = true;
        try {
          if (!refreshTokenPromise) {
            refreshTokenPromise = onRefreshToken().finally(() => {
              refreshTokenPromise = null;
            });
          }
          const newAccessToken = await refreshTokenPromise;
          if (newAccessToken) {
            setAccessToken(newAccessToken);
            originalRequest.headers.Authorization = `Bearer ${newAccessToken}`;
            return client(originalRequest);
          }
        } catch (refreshError) {
          setAccessToken(null);
          if (onAuthError) {
            onAuthError(refreshError);
          }
          return Promise.reject(refreshError);
        }
      }
      return Promise.reject(error);
    }
  );
  return client;
}

// src/context/AuthContext.tsx
import { jsx } from "react/jsx-runtime";
var AuthContext = createContext(void 0);
var apiClient;
function AuthProvider({
  children,
  apiUrl,
  googleClientId,
  enableLocalStorage = false,
  onAuthError
}) {
  const [user, setUser] = useState(null);
  const [accessToken, setAccessTokenState] = useState(null);
  const [isLoading, setIsLoading] = useState(true);
  const setAccessToken = useCallback((token) => {
    setAccessTokenState(token);
    if (enableLocalStorage) {
      if (token) {
        localStorage.setItem("accessToken", token);
      } else {
        localStorage.removeItem("accessToken");
        localStorage.removeItem("user");
      }
    }
  }, [enableLocalStorage]);
  const getAccessToken = useCallback(() => {
    if (accessToken) return accessToken;
    if (enableLocalStorage) {
      return localStorage.getItem("accessToken");
    }
    return null;
  }, [accessToken, enableLocalStorage]);
  const refreshToken = useCallback(async () => {
    try {
      const response = await fetch(`${apiUrl}/refresh`, {
        method: "POST",
        credentials: "include",
        headers: {
          "Content-Type": "application/json"
        }
      });
      if (!response.ok) {
        throw new Error("Failed to refresh token");
      }
      const data = await response.json();
      if (data.accessToken) {
        setAccessToken(data.accessToken);
        return data.accessToken;
      }
      return null;
    } catch (error) {
      console.error("Token refresh failed:", error);
      return null;
    }
  }, [apiUrl, setAccessToken]);
  useEffect(() => {
    apiClient = createApiClient(
      apiUrl,
      getAccessToken,
      setAccessToken,
      refreshToken,
      onAuthError
    );
  }, [apiUrl, getAccessToken, setAccessToken, refreshToken, onAuthError]);
  const login = useCallback((token, userData) => {
    setAccessToken(token);
    setUser(userData);
    if (enableLocalStorage) {
      localStorage.setItem("user", JSON.stringify(userData));
    }
  }, [enableLocalStorage, setAccessToken]);
  const logout = useCallback(async () => {
    try {
      await apiClient.post("/logout");
    } catch (error) {
      console.error("Logout error:", error);
    } finally {
      setAccessToken(null);
      setUser(null);
    }
  }, [setAccessToken]);
  useEffect(() => {
    const initAuth = async () => {
      try {
        const token = enableLocalStorage ? localStorage.getItem("accessToken") : null;
        const storedUser = enableLocalStorage ? localStorage.getItem("user") : null;
        if (token && storedUser) {
          setAccessToken(token);
          setUser(JSON.parse(storedUser));
          try {
            const response = await fetch(`${apiUrl}/me`, {
              headers: {
                Authorization: `Bearer ${token}`
              },
              credentials: "include"
            });
            if (response.ok) {
              const data = await response.json();
              setUser(data.user);
            } else {
              const newToken = await refreshToken();
              if (newToken) {
                const retryResponse = await fetch(`${apiUrl}/me`, {
                  headers: {
                    Authorization: `Bearer ${newToken}`
                  },
                  credentials: "include"
                });
                if (retryResponse.ok) {
                  const data = await retryResponse.json();
                  setUser(data.user);
                } else {
                  setAccessToken(null);
                  setUser(null);
                }
              } else {
                setAccessToken(null);
                setUser(null);
              }
            }
          } catch (error) {
            console.error("Failed to verify auth:", error);
            setAccessToken(null);
            setUser(null);
          }
        }
      } catch (error) {
        console.error("Auth initialization error:", error);
      } finally {
        setIsLoading(false);
      }
    };
    initAuth();
  }, [apiUrl, enableLocalStorage, refreshToken, setAccessToken]);
  const contextValue = useMemo(() => ({
    user,
    isAuthenticated: !!user,
    isLoading,
    accessToken,
    login,
    logout,
    refreshToken
  }), [user, isLoading, accessToken, login, logout, refreshToken]);
  return /* @__PURE__ */ jsx(GoogleOAuthProvider, { clientId: googleClientId, children: /* @__PURE__ */ jsx(AuthContext.Provider, { value: contextValue, children }) });
}

// src/hooks/useAuth.ts
import { useContext } from "react";
function useAuth() {
  const context = useContext(AuthContext);
  if (context === void 0) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
}

// src/components/GoogleLoginButton.tsx
import { useState as useState2 } from "react";
import { GoogleLogin } from "@react-oauth/google";
import { jsx as jsx2, jsxs } from "react/jsx-runtime";
function GoogleLoginButton({
  onSuccess,
  onError,
  text = "Continue with Google",
  className = ""
}) {
  const { login } = useAuth();
  const [isLoading, setIsLoading] = useState2(false);
  const [error, setError] = useState2(null);
  const handleGoogleLogin = async (credentialResponse) => {
    setIsLoading(true);
    setError(null);
    try {
      if (!credentialResponse.credential) {
        throw new Error("Missing Google credential");
      }
      const response = await apiClient.post("/google", {
        credential: credentialResponse.credential
      });
      if (response.data.success) {
        login(response.data.accessToken, response.data.user);
        onSuccess?.();
      } else {
        throw new Error("Login failed");
      }
    } catch (err) {
      let errorMessage = "Google login failed";
      if (err instanceof Error && err.message) {
        errorMessage = err.message;
      }
      if (typeof err === "object" && err !== null && "response" in err) {
        const response = err.response;
        errorMessage = response?.data?.error || errorMessage;
      }
      setError(errorMessage);
      onError?.(err);
      console.error("Google login error:", err);
    } finally {
      setIsLoading(false);
    }
  };
  return /* @__PURE__ */ jsxs("div", { className: `google-login-container ${className}`, children: [
    /* @__PURE__ */ jsx2(
      GoogleLogin,
      {
        onSuccess: handleGoogleLogin,
        onError: () => {
          setError("Google login failed");
          onError?.(new Error("Google login failed"));
        },
        useOneTap: false,
        text: "continue_with",
        shape: "rectangular",
        size: "large",
        width: "300"
      }
    ),
    error && /* @__PURE__ */ jsx2("div", { className: "auth-error", children: error })
  ] });
}

// src/components/EmailOtpLogin.tsx
import { useState as useState3, useCallback as useCallback2 } from "react";
import { jsx as jsx3, jsxs as jsxs2 } from "react/jsx-runtime";
function EmailOtpLogin({
  onSuccess,
  onError,
  className = ""
}) {
  const { login } = useAuth();
  const [email, setEmail] = useState3("");
  const [otp, setOtp] = useState3("");
  const [step, setStep] = useState3("email");
  const [isLoading, setIsLoading] = useState3(false);
  const [error, setError] = useState3(null);
  const [successMessage, setSuccessMessage] = useState3(null);
  const [expiresIn, setExpiresIn] = useState3(null);
  const handleSendOtp = useCallback2(async (e) => {
    e.preventDefault();
    setIsLoading(true);
    setError(null);
    setSuccessMessage(null);
    try {
      const response = await apiClient.post("/email/send-otp", { email });
      if (response.data.success) {
        setStep("otp");
        setExpiresIn(response.data.expiresIn);
        setSuccessMessage("OTP sent to your email!");
      } else {
        throw new Error("Failed to send OTP");
      }
    } catch (err) {
      let errorMessage = "Failed to send OTP";
      if (err instanceof Error && err.message) {
        errorMessage = err.message;
      }
      if (typeof err === "object" && err !== null && "response" in err) {
        const response = err.response;
        errorMessage = response?.data?.error || errorMessage;
      }
      setError(errorMessage);
      onError?.(err);
      console.error("Send OTP error:", err);
    } finally {
      setIsLoading(false);
    }
  }, [email, onError]);
  const handleVerifyOtp = useCallback2(async (e) => {
    e.preventDefault();
    setIsLoading(true);
    setError(null);
    try {
      const response = await apiClient.post("/email/verify-otp", {
        email,
        otp
      });
      if (response.data.success) {
        login(response.data.accessToken, response.data.user);
        onSuccess?.();
      } else {
        throw new Error("Invalid OTP");
      }
    } catch (err) {
      let errorMessage = "Invalid OTP";
      if (err instanceof Error && err.message) {
        errorMessage = err.message;
      }
      if (typeof err === "object" && err !== null && "response" in err) {
        const response = err.response;
        errorMessage = response?.data?.error || errorMessage;
      }
      setError(errorMessage);
      onError?.(err);
      console.error("Verify OTP error:", err);
    } finally {
      setIsLoading(false);
    }
  }, [email, otp, login, onSuccess, onError]);
  const handleBackToEmail = useCallback2(() => {
    setStep("email");
    setOtp("");
    setError(null);
    setSuccessMessage(null);
  }, []);
  return /* @__PURE__ */ jsx3("div", { className: `email-otp-login ${className}`, children: step === "email" ? /* @__PURE__ */ jsxs2("form", { onSubmit: handleSendOtp, className: "auth-form", children: [
    /* @__PURE__ */ jsxs2("div", { className: "form-group", children: [
      /* @__PURE__ */ jsx3("label", { htmlFor: "email", className: "form-label", children: "Email Address" }),
      /* @__PURE__ */ jsx3(
        "input",
        {
          id: "email",
          type: "email",
          value: email,
          onChange: (e) => setEmail(e.target.value),
          placeholder: "Enter your email",
          required: true,
          disabled: isLoading,
          className: "form-input"
        }
      )
    ] }),
    error && /* @__PURE__ */ jsx3("div", { className: "auth-error", children: error }),
    /* @__PURE__ */ jsx3(
      "button",
      {
        type: "submit",
        disabled: isLoading || !email,
        className: "auth-button",
        children: isLoading ? "Sending..." : "Send OTP"
      }
    )
  ] }) : /* @__PURE__ */ jsxs2("form", { onSubmit: handleVerifyOtp, className: "auth-form", children: [
    /* @__PURE__ */ jsxs2("div", { className: "form-group", children: [
      /* @__PURE__ */ jsx3("label", { htmlFor: "otp", className: "form-label", children: "Enter OTP" }),
      /* @__PURE__ */ jsxs2("p", { className: "form-hint", children: [
        "We sent a 6-digit code to ",
        email,
        expiresIn && ` (expires in ${Math.floor(expiresIn / 60)} minutes)`
      ] }),
      /* @__PURE__ */ jsx3(
        "input",
        {
          id: "otp",
          type: "text",
          value: otp,
          onChange: (e) => setOtp(e.target.value.replace(/\D/g, "").slice(0, 6)),
          placeholder: "123456",
          required: true,
          maxLength: 6,
          disabled: isLoading,
          className: "form-input otp-input",
          autoComplete: "one-time-code"
        }
      )
    ] }),
    successMessage && /* @__PURE__ */ jsx3("div", { className: "auth-success", children: successMessage }),
    error && /* @__PURE__ */ jsx3("div", { className: "auth-error", children: error }),
    /* @__PURE__ */ jsx3(
      "button",
      {
        type: "submit",
        disabled: isLoading || otp.length !== 6,
        className: "auth-button",
        children: isLoading ? "Verifying..." : "Verify OTP"
      }
    ),
    /* @__PURE__ */ jsx3(
      "button",
      {
        type: "button",
        onClick: handleBackToEmail,
        disabled: isLoading,
        className: "auth-button-secondary",
        children: "Change Email"
      }
    )
  ] }) });
}
export {
  AuthProvider,
  EmailOtpLogin,
  GoogleLoginButton,
  apiClient,
  useAuth
};
