// atozas-react-auth-kit/dist/index.cjs
"use client";
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  AuthProvider: () => AuthProvider,
  EmailOtpLogin: () => EmailOtpLogin,
  GoogleLoginButton: () => GoogleLoginButton,
  apiClient: () => apiClient,
  useAuth: () => useAuth
});
module.exports = __toCommonJS(index_exports);

// src/context/AuthContext.tsx
var import_react = require("react");
var import_google = require("@react-oauth/google");

// src/api/client.ts
var import_axios = __toESM(require("axios"));
var refreshTokenPromise = null;
function createApiClient(baseURL, getAccessToken, setAccessToken, onRefreshToken, onAuthError) {
  const client = import_axios.default.create({
    baseURL,
    withCredentials: true,
    // Important for cookies
    headers: {
      "Content-Type": "application/json"
    }
  });
  client.interceptors.request.use(
    (config) => {
      const token = getAccessToken();
      if (token) {
        config.headers.Authorization = `Bearer ${token}`;
      }
      return config;
    },
    (error) => Promise.reject(error)
  );
  client.interceptors.response.use(
    (response) => response,
    async (error) => {
      const originalRequest = error.config;
      if (error.response?.status === 401 && !originalRequest._retry) {
        originalRequest._retry = true;
        try {
          if (!refreshTokenPromise) {
            refreshTokenPromise = onRefreshToken().finally(() => {
              refreshTokenPromise = null;
            });
          }
          const newAccessToken = await refreshTokenPromise;
          if (newAccessToken) {
            setAccessToken(newAccessToken);
            originalRequest.headers.Authorization = `Bearer ${newAccessToken}`;
            return client(originalRequest);
          }
        } catch (refreshError) {
          setAccessToken(null);
          if (onAuthError) {
            onAuthError(refreshError);
          }
          return Promise.reject(refreshError);
        }
      }
      return Promise.reject(error);
    }
  );
  return client;
}

// src/context/AuthContext.tsx
var import_jsx_runtime = require("react/jsx-runtime");
var AuthContext = (0, import_react.createContext)(void 0);
var apiClient;
function AuthProvider({
  children,
  apiUrl,
  googleClientId,
  enableLocalStorage = false,
  onAuthError
}) {
  const [user, setUser] = (0, import_react.useState)(null);
  const [accessToken, setAccessTokenState] = (0, import_react.useState)(null);
  const [isLoading, setIsLoading] = (0, import_react.useState)(true);
  const setAccessToken = (0, import_react.useCallback)((token) => {
    setAccessTokenState(token);
    if (enableLocalStorage) {
      if (token) {
        localStorage.setItem("accessToken", token);
      } else {
        localStorage.removeItem("accessToken");
        localStorage.removeItem("user");
      }
    }
  }, [enableLocalStorage]);
  const getAccessToken = (0, import_react.useCallback)(() => {
    if (accessToken) return accessToken;
    if (enableLocalStorage) {
      return localStorage.getItem("accessToken");
    }
    return null;
  }, [accessToken, enableLocalStorage]);
  const refreshToken = (0, import_react.useCallback)(async () => {
    try {
      const response = await fetch(`${apiUrl}/refresh`, {
        method: "POST",
        credentials: "include",
        headers: {
          "Content-Type": "application/json"
        }
      });
      if (!response.ok) {
        throw new Error("Failed to refresh token");
      }
      const data = await response.json();
      if (data.accessToken) {
        setAccessToken(data.accessToken);
        return data.accessToken;
      }
      return null;
    } catch (error) {
      console.error("Token refresh failed:", error);
      return null;
    }
  }, [apiUrl, setAccessToken]);
  (0, import_react.useEffect)(() => {
    apiClient = createApiClient(
      apiUrl,
      getAccessToken,
      setAccessToken,
      refreshToken,
      onAuthError
    );
  }, [apiUrl, getAccessToken, setAccessToken, refreshToken, onAuthError]);
  const login = (0, import_react.useCallback)((token, userData) => {
    setAccessToken(token);
    setUser(userData);
    if (enableLocalStorage) {
      localStorage.setItem("user", JSON.stringify(userData));
    }
  }, [enableLocalStorage, setAccessToken]);
  const logout = (0, import_react.useCallback)(async () => {
    try {
      await apiClient.post("/logout");
    } catch (error) {
      console.error("Logout error:", error);
    } finally {
      setAccessToken(null);
      setUser(null);
    }
  }, [setAccessToken]);
  (0, import_react.useEffect)(() => {
    const initAuth = async () => {
      try {
        const token = enableLocalStorage ? localStorage.getItem("accessToken") : null;
        const storedUser = enableLocalStorage ? localStorage.getItem("user") : null;
        if (token && storedUser) {
          setAccessToken(token);
          setUser(JSON.parse(storedUser));
          try {
            const response = await fetch(`${apiUrl}/me`, {
              headers: {
                Authorization: `Bearer ${token}`
              },
              credentials: "include"
            });
            if (response.ok) {
              const data = await response.json();
              setUser(data.user);
            } else {
              const newToken = await refreshToken();
              if (newToken) {
                const retryResponse = await fetch(`${apiUrl}/me`, {
                  headers: {
                    Authorization: `Bearer ${newToken}`
                  },
                  credentials: "include"
                });
                if (retryResponse.ok) {
                  const data = await retryResponse.json();
                  setUser(data.user);
                } else {
                  setAccessToken(null);
                  setUser(null);
                }
              } else {
                setAccessToken(null);
                setUser(null);
              }
            }
          } catch (error) {
            console.error("Failed to verify auth:", error);
            setAccessToken(null);
            setUser(null);
          }
        }
      } catch (error) {
        console.error("Auth initialization error:", error);
      } finally {
        setIsLoading(false);
      }
    };
    initAuth();
  }, [apiUrl, enableLocalStorage, refreshToken, setAccessToken]);
  const contextValue = (0, import_react.useMemo)(() => ({
    user,
    isAuthenticated: !!user,
    isLoading,
    accessToken,
    login,
    logout,
    refreshToken
  }), [user, isLoading, accessToken, login, logout, refreshToken]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_google.GoogleOAuthProvider, { clientId: googleClientId, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(AuthContext.Provider, { value: contextValue, children }) });
}

// src/hooks/useAuth.ts
var import_react2 = require("react");
function useAuth() {
  const context = (0, import_react2.useContext)(AuthContext);
  if (context === void 0) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
}

// src/components/GoogleLoginButton.tsx
var import_react3 = require("react");
var import_google2 = require("@react-oauth/google");
var import_jsx_runtime2 = require("react/jsx-runtime");
function GoogleLoginButton({
  onSuccess,
  onError,
  text = "Continue with Google",
  className = ""
}) {
  const { login } = useAuth();
  const [isLoading, setIsLoading] = (0, import_react3.useState)(false);
  const [error, setError] = (0, import_react3.useState)(null);
  const handleGoogleLogin = async (credentialResponse) => {
    setIsLoading(true);
    setError(null);
    try {
      if (!credentialResponse.credential) {
        throw new Error("Missing Google credential");
      }
      const response = await apiClient.post("/google", {
        credential: credentialResponse.credential
      });
      if (response.data.success) {
        login(response.data.accessToken, response.data.user);
        onSuccess?.();
      } else {
        throw new Error("Login failed");
      }
    } catch (err) {
      let errorMessage = "Google login failed";
      if (err instanceof Error && err.message) {
        errorMessage = err.message;
      }
      if (typeof err === "object" && err !== null && "response" in err) {
        const response = err.response;
        errorMessage = response?.data?.error || errorMessage;
      }
      setError(errorMessage);
      onError?.(err);
      console.error("Google login error:", err);
    } finally {
      setIsLoading(false);
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsxs)("div", { className: `google-login-container ${className}`, children: [
    /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
      import_google2.GoogleLogin,
      {
        onSuccess: handleGoogleLogin,
        onError: () => {
          setError("Google login failed");
          onError?.(new Error("Google login failed"));
        },
        useOneTap: false,
        text: "continue_with",
        shape: "rectangular",
        size: "large",
        width: "300"
      }
    ),
    error && /* @__PURE__ */ (0, import_jsx_runtime2.jsx)("div", { className: "auth-error", children: error })
  ] });
}

// src/components/EmailOtpLogin.tsx
var import_react4 = require("react");
var import_jsx_runtime3 = require("react/jsx-runtime");
function EmailOtpLogin({
  onSuccess,
  onError,
  className = ""
}) {
  const { login } = useAuth();
  const [email, setEmail] = (0, import_react4.useState)("");
  const [otp, setOtp] = (0, import_react4.useState)("");
  const [step, setStep] = (0, import_react4.useState)("email");
  const [isLoading, setIsLoading] = (0, import_react4.useState)(false);
  const [error, setError] = (0, import_react4.useState)(null);
  const [successMessage, setSuccessMessage] = (0, import_react4.useState)(null);
  const [expiresIn, setExpiresIn] = (0, import_react4.useState)(null);
  const handleSendOtp = (0, import_react4.useCallback)(async (e) => {
    e.preventDefault();
    setIsLoading(true);
    setError(null);
    setSuccessMessage(null);
    try {
      const response = await apiClient.post("/email/send-otp", { email });
      if (response.data.success) {
        setStep("otp");
        setExpiresIn(response.data.expiresIn);
        setSuccessMessage("OTP sent to your email!");
      } else {
        throw new Error("Failed to send OTP");
      }
    } catch (err) {
      let errorMessage = "Failed to send OTP";
      if (err instanceof Error && err.message) {
        errorMessage = err.message;
      }
      if (typeof err === "object" && err !== null && "response" in err) {
        const response = err.response;
        errorMessage = response?.data?.error || errorMessage;
      }
      setError(errorMessage);
      onError?.(err);
      console.error("Send OTP error:", err);
    } finally {
      setIsLoading(false);
    }
  }, [email, onError]);
  const handleVerifyOtp = (0, import_react4.useCallback)(async (e) => {
    e.preventDefault();
    setIsLoading(true);
    setError(null);
    try {
      const response = await apiClient.post("/email/verify-otp", {
        email,
        otp
      });
      if (response.data.success) {
        login(response.data.accessToken, response.data.user);
        onSuccess?.();
      } else {
        throw new Error("Invalid OTP");
      }
    } catch (err) {
      let errorMessage = "Invalid OTP";
      if (err instanceof Error && err.message) {
        errorMessage = err.message;
      }
      if (typeof err === "object" && err !== null && "response" in err) {
        const response = err.response;
        errorMessage = response?.data?.error || errorMessage;
      }
      setError(errorMessage);
      onError?.(err);
      console.error("Verify OTP error:", err);
    } finally {
      setIsLoading(false);
    }
  }, [email, otp, login, onSuccess, onError]);
  const handleBackToEmail = (0, import_react4.useCallback)(() => {
    setStep("email");
    setOtp("");
    setError(null);
    setSuccessMessage(null);
  }, []);
  return /* @__PURE__ */ (0, import_jsx_runtime3.jsx)("div", { className: `email-otp-login ${className}`, children: step === "email" ? /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)("form", { onSubmit: handleSendOtp, className: "auth-form", children: [
    /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)("div", { className: "form-group", children: [
      /* @__PURE__ */ (0, import_jsx_runtime3.jsx)("label", { htmlFor: "email", className: "form-label", children: "Email Address" }),
      /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
        "input",
        {
          id: "email",
          type: "email",
          value: email,
          onChange: (e) => setEmail(e.target.value),
          placeholder: "Enter your email",
          required: true,
          disabled: isLoading,
          className: "form-input"
        }
      )
    ] }),
    error && /* @__PURE__ */ (0, import_jsx_runtime3.jsx)("div", { className: "auth-error", children: error }),
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      "button",
      {
        type: "submit",
        disabled: isLoading || !email,
        className: "auth-button",
        children: isLoading ? "Sending..." : "Send OTP"
      }
    )
  ] }) : /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)("form", { onSubmit: handleVerifyOtp, className: "auth-form", children: [
    /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)("div", { className: "form-group", children: [
      /* @__PURE__ */ (0, import_jsx_runtime3.jsx)("label", { htmlFor: "otp", className: "form-label", children: "Enter OTP" }),
      /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)("p", { className: "form-hint", children: [
        "We sent a 6-digit code to ",
        email,
        expiresIn && ` (expires in ${Math.floor(expiresIn / 60)} minutes)`
      ] }),
      /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
        "input",
        {
          id: "otp",
          type: "text",
          value: otp,
          onChange: (e) => setOtp(e.target.value.replace(/\D/g, "").slice(0, 6)),
          placeholder: "123456",
          required: true,
          maxLength: 6,
          disabled: isLoading,
          className: "form-input otp-input",
          autoComplete: "one-time-code"
        }
      )
    ] }),
    successMessage && /* @__PURE__ */ (0, import_jsx_runtime3.jsx)("div", { className: "auth-success", children: successMessage }),
    error && /* @__PURE__ */ (0, import_jsx_runtime3.jsx)("div", { className: "auth-error", children: error }),
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      "button",
      {
        type: "submit",
        disabled: isLoading || otp.length !== 6,
        className: "auth-button",
        children: isLoading ? "Verifying..." : "Verify OTP"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      "button",
      {
        type: "button",
        onClick: handleBackToEmail,
        disabled: isLoading,
        className: "auth-button-secondary",
        children: "Change Email"
      }
    )
  ] }) });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AuthProvider,
  EmailOtpLogin,
  GoogleLoginButton,
  apiClient,
  useAuth
});
