// atozas-auth-kit-express/src/middleware/requireAuth.ts
import { Response, NextFunction } from 'express';
import { AuthRequest, AuthConfig } from '../types';
import { verifyAccessToken } from '../utils/tokens';
import User from '../models/User';

export function createRequireAuth(config: AuthConfig) {
  return async (req: AuthRequest, res: Response, next: NextFunction) => {
    try {
      const authHeader = req.headers.authorization;
      
      if (!authHeader || !authHeader.startsWith('Bearer ')) {
        return res.status(401).json({ error: 'No token provided' });
      }
      
      const token = authHeader.substring(7);
      
      try {
        const decoded = verifyAccessToken(token, config.jwtSecret);
        
        // Fetch user from database
        const user = await User.findById(decoded.userId);
        
        if (!user) {
          return res.status(401).json({ error: 'User not found' });
        }
        
        req.user = {
          id: user._id.toString(),
          email: user.email,
          name: user.name,
          picture: user.picture,
          provider: user.provider,
        };
        
        next();
      } catch (error) {
        return res.status(401).json({ error: 'Invalid or expired token' });
      }
    } catch (error) {
      return res.status(500).json({ error: 'Internal server error' });
    }
  };
}
