// atozas-auth-kit-express/src/types.ts
import { Request } from 'express';

export interface AuthConfig {
  // Google OAuth
  googleClientId: string;
  
  // JWT Secrets
  jwtSecret: string;
  jwtRefreshSecret: string;
  
  // Token Expiry
  accessTokenExpiry?: string; // default: '15m'
  refreshTokenExpiry?: string; // default: '30d'
  
  // OTP Configuration
  otpExpiry?: number; // in minutes, default: 5
  otpLength?: number; // default: 6
  
  // Rate Limiting
  otpRateLimit?: {
    maxAttempts: number; // default: 3
    windowMs: number; // default: 15 * 60 * 1000 (15 minutes)
  };
  
  // Email Adapter
  sendEmailOtp: (to: string, otp: string) => Promise<void>;
  
  // Optional: Custom MongoDB connection
  mongooseConnection?: any;
  
  // Cookie options
  cookieOptions?: {
    httpOnly?: boolean;
    secure?: boolean;
    sameSite?: 'strict' | 'lax' | 'none';
    domain?: string;
    path?: string;
  };
  
  // Frontend URL for CORS
  frontendUrl?: string;
}

export interface UserPayload {
  id: string;
  email: string;
  name?: string;
  picture?: string;
  provider: 'google' | 'email';
}

export interface AuthRequest extends Request {
  user?: UserPayload;
}

export interface TokenPayload {
  userId: string;
  email: string;
  provider: 'google' | 'email';
}

export interface OtpDocument {
  email: string;
  otp: string;
  expiresAt: Date;
  used: boolean;
  ipAddress?: string;
  createdAt: Date;
}

export interface UserDocument {
  email: string;
  name?: string;
  picture?: string;
  provider: 'google' | 'email';
  googleId?: string;
  createdAt: Date;
  updatedAt: Date;
}
