// atozas-react-auth-kit/src/components/GoogleLoginButton.tsx
import React, { useState } from 'react';
import { GoogleLogin, CredentialResponse } from '@react-oauth/google';
import { useAuth } from '../hooks/useAuth';
import { apiClient } from '../context/AuthContext';

interface GoogleLoginButtonProps {
  onSuccess?: () => void;
  onError?: (error: unknown) => void;
  text?: string;
  className?: string;
}

export function GoogleLoginButton({
  onSuccess,
  onError,
  text = 'Continue with Google',
  className = '',
}: GoogleLoginButtonProps) {
  const { login } = useAuth();
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);

  const handleGoogleLogin = async (credentialResponse: CredentialResponse) => {
    setIsLoading(true);
    setError(null);

    try {
      if (!credentialResponse.credential) {
        throw new Error('Missing Google credential');
      }

      const response = await apiClient.post('/google', {
        credential: credentialResponse.credential,
      });

      if (response.data.success) {
        login(response.data.accessToken, response.data.user);
        onSuccess?.();
      } else {
        throw new Error('Login failed');
      }
    } catch (err: unknown) {
      let errorMessage = 'Google login failed';
      if (err instanceof Error && err.message) {
        errorMessage = err.message;
      }
      if (typeof err === 'object' && err !== null && 'response' in err) {
        const response = (err as { response?: { data?: { error?: string } } }).response;
        errorMessage = response?.data?.error || errorMessage;
      }
      setError(errorMessage);
      onError?.(err);
      console.error('Google login error:', err);
    } finally {
      setIsLoading(false);
    }
  };

  return (
    <div className={`google-login-container ${className}`}>
      <GoogleLogin
        onSuccess={handleGoogleLogin}
        onError={() => {
          setError('Google login failed');
          onError?.(new Error('Google login failed'));
        }}
        useOneTap={false}
        text="continue_with"
        shape="rectangular"
        size="large"
        width="300"
      />
      {error && <div className="auth-error">{error}</div>}
    </div>
  );
}
